pytensor/tensor/rewriting/linalg.py

```diff
@@ -377,3 +383,96 @@ def local_lift_through_linalg(
                 return [block_diag(*inner_matrices)]
             else:
                 raise NotImplementedError  # pragma: no cover
+
+
+# Det Diag Rewrites
+@register_canonicalize
+@register_stabilize
+@node_rewriter([det])
+def rewrite_det_diag_from_eye_mul(fgraph, node):
+    """
+     This rewrite takes advantage of the fact that for a diagonal matrix, the determinant value is the product of its diagonal elements.
+
+    The presence of a diagonal matrix is detected by inspecting the graph. This rewrite can identify diagonal matrices that arise as the result of elementwise multiplication with an identity matrix. Specialized computation is used to make this rewrite as efficient as possible, depending on whether the multiplication was with a scalar, vector or a matrix.
+
+    Parameters
+    ----------
+    fgraph: FunctionGraph
+        Function graph being optimized
+    node: Apply
+        Node of the function graph to be optimized
+
+    Returns
+    -------
+    list of Variable, optional
+        List of optimized variables, or None if no optimization was performed
+    """
+    node_input = node.inputs[0]
+    # Check if the op is Elemwise and mul
+    if not (
+        node_input.owner is not None
+        and isinstance(node_input.owner.op, Elemwise)
+        and isinstance(node_input.owner.op.scalar_op, Mul)
+    ):
+        return None
+
+    # Find whether any of the inputs to mul is Eye
+    inputs_to_mul = node_input.owner.inputs
+    eye_input = [
+        mul_input
+        for mul_input in inputs_to_mul
+        if mul_input.owner and isinstance(mul_input.owner.op, Eye)
+    ]
+    if eye_input and eye_input[0].broadcastable[-2:] != (False, False):
+        return None
+
+    # Get all non Eye inputs (scalars/matrices/vectors)
+    non_eye_inputs = list(set(inputs_to_mul) - set(eye_input))
+
+    # Dealing with only one other input
+    if len(non_eye_inputs) >= 2:
+        return None
+
+    # Rewrite is only applied if all the shapes are known
+    if non_eye_inputs[0].type.shape[-2:] == (None, None) or eye_input[0].type.shape[
+        -2:
+    ] == (None, None):
+        return None
+    # This ensures that the rewrite is NOT applied to cases of degenerate eye (pt.eye(1)) and rectangle eye (pt.eye(7,5))
```

Comment:
This comment should be on line 436 I think? But it's also not true, the broadcast pattern check covers the degenerate case. 

This check is really strict, and will miss a lot of chances to rewrite because we're being so cautious. I think it might be good to instead insert an `Assert` `Op` if any of the shapes are unknown. I know @ricardoV94 doesn't like them, but this seems like a good use case.

Reply:
So should i define a new assert statement for unknown shapes and add those here instead of using the if condition?

Reply:
You should only have to look at the broadcastable pattern, doesn't matter if shape is known or a specific number

Reply:
Det on a non-square matrix will raise an error at runtime.

Consider this graph:

```
x = pt.tensor('x', shape=(10, 3))
eye = pt.eye(10, 3)
z = x * eye
out = pt.linalg.det(z)
```

This should error, but if we only check the broadcast pattern of `eye` and `x`, we will incorrectly rewrite this to `pt.diagonal(x).prod()`, which would be something like `det(x[:3, :])`. The runtime error raised by the `Det` op would not be raised (since it was rewritten away), and we would introduce a silent numerical error into the user's graph.

My proposal for using an Assert was basically to preserve the runtime error.

Reply:
I really dislike adding Asserts. The Theano philosophy with rewrites was usually (not always) to assume you have a valid graph and rewrite accordingly. We added the tag `shape_unsafe` to mark rewrites that rely on assuming valid shapes. Users can disable it if they want to debug weird results that may be due to shape bugs.

https://github.com/pymc-devs/pytensor/pull/381